feat: add an AI page (#821)
scientific-python/cookie: docs/guides/ai.md

It helps to separate two very different things that often get lumped together
as "Agentic AI":

- A developer driving an interactive AI harness with a capable model, reading
  the output, and taking responsibility for the result. This is a power tool,
  much like an editor or a linter.
- Low-cost models running unattended in automated systems that mass-produce
  pull requests. This is what most people mean by "AI slop", and it is the
  source of most of the frustration maintainers feel about AI contributions.

The recommendations below are aimed at the first case, and at keeping your
project from being overwhelmed by the second.

:::{note}
The first point does hide something: the tool depends on the developer guiding
it (just like any other tool). You will also see users with very little coding
experience using these tools to produce low quality contributions. How someone
learns to code in this new era is still something unsolved.

If you maintain a project: Try to engage with the human. If they are willing to
interact (and not just type "address the review" into their harness), treat
them like a human, even if you also see the AI working on their behalf. They

> also may use AI to address a language barrier.

Reviewer: I would expand this to be inclusive beyond just 'language barrier'. Modern Ai can be a powerful accessibility. It can be a competent translator and break down the language barrier, and it can read-aloud for the visually impaired, parse large texts for people with dyslexia, etc.

```suggestion
also may use AI to address a language barrier, or accommodate a disability.
```

Author: Great point!

```

:::
:::{tab-item} Minimal

AI-generated PRs are discouraged or restricted. Use this if you have limited
review capacity.

```markdown
# AI Policy

We do not accept unsolicited AI-generated pull requests. Please open an issue
to discuss before contributing. Fully-reviewed, disclosed AI-assisted fixes may
be considered case by case.
```

:::
::::

## `AGENTS.md`

Harnesses read a project context file to learn how your repository works --
preferred command runners, architecture notes, conventions, and gotchas. A good
context file makes the AI far more effective without bloating every prompt. The
cross-tool standard is [`AGENTS.md`][agents-md]; most harnesses can generate a
first draft for you (often via an `/init` command).

Keep it focused on what is *not* obvious from the code: how to run the tests,
which tools to prefer, where generated files live, and any traps. Treat it as
documentation you maintain, not a dumping ground.

:::{note} Claude Code and `AGENTS.md`

Claude Code is the only major harness to *not* read `AGENTS.md`. You can support
both with a symlink, keeping a single source of truth:

```bash
ln -s AGENTS.md CLAUDE.md
```

> You can also mention `@AGENTS.md` inside `CLAUDE.md` if you want to add
> specific instructions; this is true for all the other harnesses too
> (`copilot-instructions.md`, etc).

Reviewer: the @file-name.md is a special feature of ClaudeCode.  Like a template, it injects the full text of the target file into context... yet the ClaudeCode best practices shows the example "See @README.md for project overview and @package.json for available npm commands." which would create a garbled mess...

Its nice that this makes CLAUDE.md composable, but its not a standard.

I think the wording is a little misleading here, because @file-injection is not true for all harnesses.

```suggestion
You can also mention `@AGENTS.md` inside `CLAUDE.md` which injects the text form the @/path/to/file.md into CLAUDE.md when ClaudeCode reads it.  You can include ClaudeCode specific instructions, and inject text from any other .md file you wish `copilot-instructions.md` for example.
```

Author: "All" is incorrect, but most of them do have this feature - I think every single one I've tried (OpenCode, Pi, Codex, Copilot). If it's not part of the agent, most models should realize you are referring to a file and try to read it anyway. Also, it doesn't inject, it reads and sends a structured message. So it's perfectly fine to add in inside a sentence. It will also know what file it referenced.

Author: It's wrong about copilot below, that's also `@`, that's where I first used it. Probably confusing another of [Microsoft's 75 tools named copilot](https://teybannerman.com/strategy/2026/03/31/how-many-microsoft-copilot-are-there.html).

"eager" means it reads the file and sends it along as text, vs. attaching the file and giving the model options about reading it.

:robot: _AI text below_ :robot:

Tool | Syntax | Notes
-- | -- | --
Claude Code | @file | Eager, full-file inline
pi | @file | Eager; works in TUI and on command line
Cursor | @Files (+ @Folders, @Code, @Git, @Docs, @Codebase) | The tool that popularized this; richest taxonomy
Windsurf (Cascade) | @file (+ @function, @docs) | Explicitly modeled on Cursor
Cline | @/path (+ @problems, @terminal, @git, @url) | Eager, full-file
Gemini CLI | @file | Eager, client-side injection
opencode | @file (+ @alias references) | On-demand read
Codex | @file / /mention | Attaches to context; completeness less guaranteed
Copilot | #file | @ reserved for participants (@workspace)
Aider | /add (slash command) | Different paradigm entirely

Reviewer: very cool, last I checked (more than 15 minutes ago) it was just a ClaudeCode thing, sounds like its becoming standard(ish) in AGENTS.md-type-files

:::

How you track the file is a separate decision:

::::{tab-set}
:::{tab-item} Commit it

Commit `AGENTS.md` so every contributor (and their harness) shares the same
project context. This is a good default for projects with at least one

> maintainer also using AI harnesses. (Ignoring `CLAUDE.md` and `.claude/` is

Reviewer 2: ```suggestion
maintainer also using AI harnesses. (Ignoring `CLAUDE.md` and `.claude/` in your `.gitignore` is
```

> maintainer also using AI harnesses. (Ignoring `CLAUDE.md` and `.claude/` in your `.gitignore` is

Reviewer: `CLAUDE.local.md` should also always be ignored and never committed. It is designed for per-user personal settings.

Projects that wish to totally exclude CLAUDE.md could direct users to just @AGENTS.md in their CLAUDE.local.md as a good practice.

Author: I didn't know about that one!

  process - even in hidden comments. When you point an agent at outside
  material (e.g. "triage these issues" or a CI run URL), review what it does
  rather than letting it act unattended, and don't combine untrusted input with
  destructive or credentialed access. This is unfortunately a big issue with
  setting up an automated issue processing system.

## What AI is good at

AI is fantastic at anything that has a clear pass/fail condition. This means
it's great at fixing up a failing PR, addressing linter failures, polishing off
anything that's failing tests into making it pass tests. That's why good tests
and strong linters and type checking are so helpful to AI, they give it a
better pass/fail to work with. Do keep an eye on it, though, sometimes it will
skip something instead of fixing it; sometimes that's correct, but decision
making is not as strong of an AI skill as pass/fail checks!

AI knows a massive library of tricks and details. It will hallucinate ones
sometimes, of course (that's why the pass/fail is important above!). Make it
validate anything (newer models often have this in the system prompts, so it is

> model and harness dependent - for example, Claude Opus 4.8+ is paranoid

Reviewer 2: ```suggestion
model and harness dependent - for example, Claude Opus 4.8+ acts paranoid
```

Reviewer 2: Just to clarify, even though I am a huge proponent of using AI tools to help with coding, I still think it's helpful to avoid personification of AI in the phrasing. (Admittedly, I still slip, and tell my agents, "bro don't do this!", so it's really tempting haha.)

Author: Haha, great point. :)
